verification/verify.go

```diff
@@ -153,3 +160,68 @@ func IsRetryable(err error) bool {
 	_, retryable := err.(Retryable)
 	return retryable
 }
+
+func (sv *SegmentVerifier) countPixelParams(params *Params) (*Results, error) {
+
+	if params.Orchestrator == nil {
```

Comment:
I think we may need to keep it.  With it removed by the following patch:
```diff
diff --git a/verification/verify.go b/verification/verify.go
index 6a429060..9a10b82b 100644
--- a/verification/verify.go
+++ b/verification/verify.go
@@ -165,10 +165,6 @@ func IsRetryable(err error) bool {
 
 func (sv *SegmentVerifier) countPixelParams(params *Params) ([]int64, error) {
 
-       if params.Orchestrator == nil {
-               return nil, ErrPixelsAbsent
-       }
-
        if len(params.Results.Segments) != len(params.Renditions) {
                return nil, ErrPixelMismatch
        }
```
many tests fail:
```
$ go test
E0211 16:05:55.057727   31411 epic.go:151] Could not submit request Post : unsupported protocol scheme ""
--- FAIL: TestVerify (1.33s)
    verify_test.go:62: 
        	Error Trace:	verify_test.go:62
        	Error:      	Not equal: 
        	            	expected: *errors.errorString(&errors.errorString{s:"PixelsAbsent"})
        	            	actual  : verification.Retryable(verification.Retryable{error:(*errors.errorString)(0xc00007d620)})
        	Test:       	TestVerify
    verify_test.go:138: 
        	Error Trace:	verify_test.go:138
        	Error:      	Not equal: 
        	            	expected: *errors.errorString(&errors.errorString{s:"PixelsAbsent"})
        	            	actual  : verification.Retryable(verification.Retryable{error:(*errors.errorString)(0xc00007d620)})
        	Test:       	TestVerify
    verify_test.go:139: 
        	Error Trace:	verify_test.go:139
        	Error:      	Not equal: 
        	            	expected: "mno"
        	            	actual  : "pqr"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-mno
        	            	+pqr
        	Test:       	TestVerify
    verify_test.go:143: 
        	Error Trace:	verify_test.go:143
        	Error:      	Not equal: 
        	            	expected: *errors.errorString(&errors.errorString{s:"PixelsAbsent"})
        	            	actual  : verification.Retryable(verification.Retryable{error:(*errors.errorString)(0xc00007d620)})
        	Test:       	TestVerify
    verify_test.go:144: 
        	Error Trace:	verify_test.go:144
        	Error:      	Not equal: 
        	            	expected: "mno"
        	            	actual  : "stu"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-mno
        	            	+stu
        	Test:       	TestVerify
demuxer: Unable to open file
Freeing input based on OPEN INPUT error
E0211 16:05:56.387836   31411 ffmpeg.go:292] Transcoder Return : No such file or directory
demuxer: Unable to open file
Freeing input based on OPEN INPUT error
E0211 16:05:56.841673   31411 ffmpeg.go:292] Transcoder Return : No such file or directory
FAIL
exit status 1
FAIL	github.com/livepeer/go-livepeer/verification	4.061s
```

Reply:
Looks like these tests were originally written to handle a scenario where the verifier does not return pixel counts so they are checking for `ErrPixelsAbsent`. Since that is no longer expected behavior as of this PR we can adjust the tests - a test that expected `ErrPixelsAbsent` should no longer expect an error since pixel counting is run within `SegmentVerifier.Verify()`.

Reply:
What about the tests that expect "mno" but actually get "pqr" or "stu" ?

Also, the comments above the tests that expected `ErrPixelsAbsent` indicate that errors are indeed expected.  So now they expect `ErrPixelMismatch` instead:

```diff
diff --git a/verification/verify.go b/verification/verify.go
index 8809badb..d0b87e7a 100644
--- a/verification/verify.go
+++ b/verification/verify.go
@@ -166,10 +166,6 @@ func IsRetryable(err error) bool {
 
 func (sv *SegmentVerifier) countPixelParams(params *Params) ([]int64, error) {
 
-       if params.Orchestrator == nil {
-               return nil, ErrPixelsAbsent
-       }
-
        if len(params.Results.Segments) != len(params.Renditions) {
                return nil, ErrPixelMismatch
        }
diff --git a/verification/verify_test.go b/verification/verify_test.go
index e23f0c83..c6ad7b00 100644
--- a/verification/verify_test.go
+++ b/verification/verify_test.go
@@ -59,7 +59,7 @@ func TestVerify(t *testing.T) {
        }}
        res, err = sv.Verify(&Params{Results: data})
        assert.Nil(res)
-       assert.Equal(ErrPixelsAbsent, err)
+       assert.Equal(ErrPixelMismatch, err)
 
        // check pixel count fails w/ external verifier
        data.Segments = append(data.Segments, &net.TranscodedSegmentData{Url: "def", Pixels: verifier.results.Pixels[1]})
@@ -135,12 +135,12 @@ func TestVerify(t *testing.T) {
        // Good score but incorrect pixel list should still fail
        verifier.results = &Results{Score: 10.0, Pixels: []int64{789}}
        res, err = sv.Verify(&Params{ManifestID: "pqr", Results: data})
-       assert.Equal(ErrPixelsAbsent, err)
+       assert.Equal(ErrPixelMismatch, err)
        assert.Equal("mno", string(res.ManifestID)) // Still return best result
        // Higher score and incorrect pixel list should prioritize pixel count
        verifier.results = &Results{Score: 20.0, Pixels: []int64{789}}
        res, err = sv.Verify(&Params{ManifestID: "stu", Results: data})
-       assert.Equal(ErrPixelsAbsent, err)
+       assert.Equal(ErrPixelMismatch, err)
        assert.Equal("mno", string(res.ManifestID)) // Still return best result
 
        // Check *not* retryable; should never get a result
```
now yeilds:
```
E0212 06:24:39.765784   11284 epic.go:151] Could not submit request Post : unsupported protocol scheme ""
--- FAIL: TestVerify (1.47s)
    verify_test.go:139: 
        	Error Trace:	verify_test.go:139
        	Error:      	Not equal: 
        	            	expected: "mno"
        	            	actual  : "pqr"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-mno
        	            	+pqr
        	Test:       	TestVerify
    verify_test.go:144: 
        	Error Trace:	verify_test.go:144
        	Error:      	Not equal: 
        	            	expected: "mno"
        	            	actual  : "stu"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-mno
        	            	+stu
        	Test:       	TestVerify
demuxer: Unable to open file
Freeing input based on OPEN INPUT error
E0212 06:24:41.236904   11284 ffmpeg.go:292] Transcoder Return : No such file or directory
demuxer: Unable to open file
Freeing input based on OPEN INPUT error
E0212 06:24:41.701484   11284 ffmpeg.go:292] Transcoder Return : No such file or directory
FAIL
exit status 1
FAIL	github.com/livepeer/go-livepeer/verification	4.329s
```

Reply:
Looks like those tests were checking that if the length of the slice of computed pixels counts is N and the length of the slice of reported pixel counts is M that the segment verifier will *not* return the `Params` object passed in despite it having a higher score than the results already tracked by the segment verifier. However, the segment verifier is returning the `Params` object passed in right now in those tests. The bug is here:

```
if len(pxls) != len(params.Results.Segments) {
    pxls, err = sv.countPixelParams(params)
}
```

In these tests, the verifier returns a slice of computed pixel counts that has an incorrect length. We previously set the error to `ErrPixelsAbsent` in this case, but now we're creating a fresh slice of computed pixel counts. Instead of counting the pixels when `len(pxls) != len(params.Results.Segments)` we can count the pixels when `len(pxls) == 0` and then after that if `len(pxls) != len(params.Results.Segments)` we set the error to `ErrPixelMismatch`.

Reply:
Thanks for explaining that.  It makes good sense, but when I apply the following diff I am getting the same exact failure in the tests:
```diff
diff --git a/verification/verify.go b/verification/verify.go
index 8809badb..c630ce81 100644
--- a/verification/verify.go
+++ b/verification/verify.go
@@ -121,9 +121,12 @@ func (sv *SegmentVerifier) Verify(params *Params) (*Params, error) {
 
        // Check pixel counts
        if (err == nil || IsRetryable(err)) && res != nil && params.Results != nil {
-               if len(pxls) != len(params.Results.Segments) {
+               if len(pxls) == 0 {
                        pxls, err = sv.countPixelParams(params)
                }
+               if len(pxls) != len(params.Results.Segments) {
+                       err = ErrPixelMismatch
+               }
                for i := 0; err == nil && i < len(params.Results.Segments) && i < len(pxls); i++ {
                        reportedPixels := params.Results.Segments[i].Pixels
                        verifiedPixels := pxls[i]
@@ -166,10 +169,6 @@ func IsRetryable(err error) bool {
 
 func (sv *SegmentVerifier) countPixelParams(params *Params) ([]int64, error) {
 
-       if params.Orchestrator == nil {
-               return nil, ErrPixelsAbsent
-       }
-
        if len(params.Results.Segments) != len(params.Renditions) {
                return nil, ErrPixelMismatch
        }
diff --git a/verification/verify_test.go b/verification/verify_test.go
index e23f0c83..c6ad7b00 100644
--- a/verification/verify_test.go
+++ b/verification/verify_test.go
@@ -59,7 +59,7 @@ func TestVerify(t *testing.T) {
        }}
        res, err = sv.Verify(&Params{Results: data})
        assert.Nil(res)
-       assert.Equal(ErrPixelsAbsent, err)
+       assert.Equal(ErrPixelMismatch, err)
 
        // check pixel count fails w/ external verifier
        data.Segments = append(data.Segments, &net.TranscodedSegmentData{Url: "def", Pixels: verifier.results.Pixels[1]})
@@ -135,12 +135,12 @@ func TestVerify(t *testing.T) {
        // Good score but incorrect pixel list should still fail
        verifier.results = &Results{Score: 10.0, Pixels: []int64{789}}
        res, err = sv.Verify(&Params{ManifestID: "pqr", Results: data})
-       assert.Equal(ErrPixelsAbsent, err)
+       assert.Equal(ErrPixelMismatch, err)
        assert.Equal("mno", string(res.ManifestID)) // Still return best result
        // Higher score and incorrect pixel list should prioritize pixel count
        verifier.results = &Results{Score: 20.0, Pixels: []int64{789}}
        res, err = sv.Verify(&Params{ManifestID: "stu", Results: data})
-       assert.Equal(ErrPixelsAbsent, err)
+       assert.Equal(ErrPixelMismatch, err)
        assert.Equal("mno", string(res.ManifestID)) // Still return best result
 
        // Check *not* retryable; should never get a result
```

Reply:
Ah now the problem is that `ErrPixelMismatch` is a `Retryable` error so this code:

```
if IsRetryable(err) {
    // Append results to sv.results
}
```

will execute and the `Params` passed into the segment verifier will end up being returned because it has the highest score despite having a reported pixels slice with an incorrect length. We could return `ErrPixelsAbsent` here for now since it is not `Retryable` and once we implement #1327 we can return `ErrPixelMismatch` since `Params` that trigger a fatal retryable error such as `ErrPixelMismatch` will not be eligible to be returned.

Reply:
I believe this is resolved by e05a63d37931c4f2c795e5debe0cf71a55fae283